pkg/operator/controller/gatewayclass/subscription.go

```diff
@@ -0,0 +1,120 @@
+package gatewayclass
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// ensureServiceMeshOperatorSubscription attempts to ensure that a subscription
+// for servicemeshoperator is present and returns a Boolean indicating whether
+// it exists, the subscription if it exists, and an error value.
+func (r *reconciler) ensureServiceMeshOperatorSubscription(ctx context.Context) (bool, *operatorsv1alpha1.Subscription, error) {
+	name := types.NamespacedName{
+		// TODO Use named consts for the namespace and name.
+		Namespace: "openshift-operators",
+		Name:      "servicemeshoperator",
+	}
+	have, current, err := r.currentSubscription(ctx, name)
+	if err != nil {
+		return false, nil, err
+	}
+
+	desired, err := desiredSubscription(name)
+	if err != nil {
+		return have, current, err
+	}
+
+	switch {
+	case !have:
+		if err := r.createSubscription(ctx, desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentSubscription(ctx, name)
+	case have:
+		if updated, err := r.updateSubscription(ctx, current, desired); err != nil {
+			return have, current, err
+		} else if updated {
+			return r.currentSubscription(ctx, name)
+		}
+	}
+	return true, current, nil
+}
+
+// desiredSubscription returns the desired subscription.
+func desiredSubscription(name types.NamespacedName) (*operatorsv1alpha1.Subscription, error) {
+	subscription := operatorsv1alpha1.Subscription{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: name.Namespace,
+			Name:      name.Name,
+		},
+		Spec: &operatorsv1alpha1.SubscriptionSpec{
+			// TODO: Use named consts if available.
+			Channel:                "stable",
+			InstallPlanApproval:    operatorsv1alpha1.ApprovalAutomatic,
+			Package:                "servicemeshoperator",
+			CatalogSource:          "redhat-operators",
+			CatalogSourceNamespace: "openshift-marketplace",
+		},
```

Comment:
It doesn't look like like anything in this Subscription would ever change even if the version changed. I wonder if there needs to be another field set in desiredSubscription, e.g. StartingCSV ?

Reply:
What is the advantage of specifying `StartingCSV` in the Subscription?  The only thing I can think of is if we need a specific minimum 2.3.z release (e.g. 2.3.2 if it has working automated deployments).